File: jli/src/expr.rs
use crate::ast;
use crate::object::Object;
use crate::token::Token;
use crate::Result;
use std::rc::Rc;

pub type VisitorResult = Result<Rc<dyn Object>>;

ast!(Expr -> VisitorResult [
    Assign(name: Token, value: Box<dyn Expr>),
    Binary(left: Box<dyn Expr>, operator: Token, right: Box<dyn Expr>),
    Call(callee: Box<dyn Expr>, paren: Token, arguments: Vec<Box<dyn Expr>>),
    Get(object: Rc<dyn Expr>, name: Token),
    Grouping(expression: Box<dyn Expr>),
    Literal(value: Rc<dyn Object>),
    Logical(left: Box<dyn Expr>, operator: Token, right: Box<dyn Expr>),
    Set(object: Rc<dyn Expr>, name: Token, value: Box<dyn Expr>),
    This(keyword: Token),
    Unary(operator: Token, right: Box<dyn Expr>),
    Variable(name: Token),
]);

File: jlox/src/dev/jona/lox/Expr.java
package dev.jona.lox;

import javax.annotation.processing.Generated;
import java.util.List;

@Generated("dev.jona.tool.GenerateAst")
interface Expr {
    <R> R accept(Visitor<R> visitor);

    interface Visitor<R> {
        R visitAssignExpr(Assign expr);
        R visitBinaryExpr(Binary expr);
        R visitCallExpr(Call expr);
        R visitGetExpr(Get expr);
        R visitGroupingExpr(Grouping expr);
        R visitLiteralExpr(Literal expr);
        R visitLogicalExpr(Logical expr);
        R visitSetExpr(Set expr);
        R visitThisExpr(This expr);
        R visitUnaryExpr(Unary expr);
        R visitVariableExpr(Variable expr);
    }

    record Assign(Token name, Expr value) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitAssignExpr(this);
        }
    }

    record Binary(Expr left, Token operator, Expr right) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitBinaryExpr(this);
        }
    }

    record Call(Expr callee, Token paren, List<Expr> arguments) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitCallExpr(this);
        }
    }

    record Get(Expr object, Token name) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitGetExpr(this);
        }
    }

    record Grouping(Expr expression) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitGroupingExpr(this);
        }
    }

    record Literal(Object value) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitLiteralExpr(this);
        }
    }

    record Logical(Expr left, Token operator, Expr right) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitLogicalExpr(this);
        }
    }

    record Set(Expr object, Token name, Expr value) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitSetExpr(this);
        }
    }

    record This(Token keyword) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitThisExpr(this);
        }
    }

    record Unary(Token operator, Expr right) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitUnaryExpr(this);
        }
    }

    record Variable(Token name) implements Expr {
        @Override
        public <R> R accept(Visitor<R> visitor) {
            return visitor.visitVariableExpr(this);
        }
    }
}
